README.md
# Kubernetes vCD operator

A python based proof of concept of an operator to manage VMware Cloud Director ressources.

* Free software: MIT license
* Documentation: [kvcd.readthedocs.io](https://kvcd.readthedocs.io).

## Features

`kvcd` is a proof-of-concept of a python based kubernetes operator to manage VMware vCloud Director resources.

With this first **preview** release, you can manage the state of **vApp** objects with the following configuration
properties:

* Parent Org (*creation only*)
* Parent Org VDC (*creation only*)
* Name (*creation only*)
* Description
* Fence mode (*creation only*)
* EULAs acceptance (*creation only*)
* Ownership
* Power status: on/off
* deploymentLease
* storageLease
* source catalog: if cloned from vCD library item
* source template: if cloned from vCD library item
* metadata: through the Kubernetes resources annotations: ReadOnly on vCloud Director side.

The operator also populates a `status.backing` dictionnary with the following properties according to the vCloud
Director data:

* UUID
* status
* vcd_vapp_href
* vcd_vdc_href
* owner
* deploymentLeaseInSeconds
* storageLeaseInSeconds
* metadata

If a deviation is detected with the declared `specs` of the object: a reconciliation is made to apply the state from
the declared `specs`.

> It may remains some use-case where the reconciliation will fail like when trying to apply a
power-on expected state on an *expired* object.*

The same reconciliation process will occurs when a change is made to the `specs` object declaration to apply changes
to the *backing* object on vCloud Director.

> Modification of some `specs` properties will be ignored, such as `org`, `vdc`, `fence_mode` or `accept_all_eulas`.

## Installation

### Stable release

To install Kubernetes vCD operator, run this command in your terminal:

```bash
pip install kvcd
```

This is the preferred method to install Kubernetes vCD operator, as it will always install the most recent stable release.

If you don't have [pip](https://pip.pypa.io) installed, this [Python installation guide](https://docs.python-guide.org/en/latest/starting/installation/) can guide you through the process.

### From sources

The sources for Kubernetes vCD operator can be downloaded from the [Github repo](https://github.com/lrivallain/kvcd).

You can clone the public repository:

```bash
git clone git://github.com/lrivallain/kvcd
```

Once you have a copy of the source, you can install it with:

```bash
pip install .
```

## Usage

### Configuration

Create a `.venv` file with the following content, according to your setup:

```bash
# Hostname of the vCloud instance | mandatory
KVCD_VCD_HOST=vcloud.domain

# HTTPS port to the vCloud instance | optional: 443 by default
KVCD_VCD_PORT=443

# Organisation of the vCloud user | mandatory
KVCD_VCD_ORG=orgX

# vCloud credentials to use as a service account | mandatory
KVCD_VCD_USERNAME=kvcd-svc
KVCD_VCD_PASSWORD=**********

# Verify the SSL certificate to connect to vCD instance | optional: yes by default
KVCD_VCD_VERIFY_SSL=yes

# Delay between two refresh of the vCD session | optional: 3600 by default
KVCD_VCD_REFRESH_SESSION_INTERVAL=3600

# Refresh interval of the vCloud instance data for each object | optional: 10 by default
KVCD_REFRESH_INTERVAL=10

# Warming up duration | optional: 30 by default
KVCD_REFRESH_INITIAL_DELAY=30

# Reduce the number of timer checks when the ressource is changed | optional: 10 by default
KVCD_REFRESH_IDLE_DELAY=10
```

### Test namespace

For the test, we will deploy a test namespace on the Kubernetes cluster:

```bash
kubectl create namespace "test-kvcd"
```

### Deploy the CRD

Deploy the Custom Resource Definition by using the following commands:

```bash
export KVCD_VERSION="v0.1.0"
kubectl apply -f https://github.com/lrivallain/kvcd/releases/download/${KVCD_VERSION}/kvcd-crds.yaml
```

This will deploy the definitions of objects that are managed by the current operator.

### Run it

The current PoC can be run locally or be embeded in a Kubernetes deployment (with pods, service...).

* The local deployment is easier to debug and troubleshoot but requires that the script remain running
to manage resources.
* The embeded deployment is probably a better choice for a real-use-case but may be more complexe to troubleshoot.

#### Run locally

From your dev machine, with access to the Kubernetes cluster, run the following:

```bash
# Run operator with kopf command and listen to specific namespaces
kopf run -m kvcd.main --namespace="test-kvcd" --verbose
```

#### Run in your K8S cluster

To run the operator in the Kubernetes cluster itself, a new namespace `kvcd-system` is used.
You can deploy the required components by running:

```bash
kubectl apply -f https://github.com/lrivallain/kvcd/releases/download/${KVCD_VERSION}/operator-deployment.yaml

# convert the local .env file to a configMap
kubectl create configmap -n kvcd-system kvcd-config --from-env-file=.env

# ensure that the expected pod is running
kubectl get pod -n kvcd-system

# checking logs
kubectl logs -n kvcd-system -f deployment/kvcd-operator
```

### Test

You can now deploy a test empty vApp to check the behavior of the operator:

```bash
cat << EOF | kubectl apply -f -
---
apiVersion: kvcd.lrivallain.dev/v1
kind: VcdVapp
metadata:
  name: kvcd-test-vapp1
  namespace: test-kvcd
  annotations:
    app: app01
    isprod: "false"
    version: "0.1.0"
spec:
  description: Test description
  org: <name of your org>
  vdc: <name of your org VDC>
  owner: <name of the owner to set>
EOF

kubectl get vcdvapp
```

After a couple of seconds, you should get something like this:

```bash
NAME              ORG        VDC             STATUS       UUID
kvcd-test-vapp1   <org>      <org vdc>       Resolved     urn:vcloud:vapp:a2871e71-49ab-48a4-a6dc-4c11743b7ba7
```

You can now edit fields values, delete or manage the vApp like a kube object.

### Cleanup

```bash
kubectl delete -n test-kvcd vcdvapp kvcd-test-vapp1
kubectl delete namespace test-kvcd
```

If you did deploy the operator in the cluster and you want to remove it:

```bash
kubectl delete ns kvcd-system
```


HISTORY.md
# History


## 0.1.0 (2021-08-13)

* First release on PyPI.

`kvcd` is a proof-of-concept of a python based kubernetes operator to manage VMware vCloud Director resources.

With this first **preview** release, you can manage the state of vApp objects with the following configuration items:

* Org (*creation only*)
* Org VDC (*creation only*)
* Name (*creation only*)
* Description
* Fence mode (*creation only*)
* EULAs acceptance (*creation only*)
* Ownership
* Power status: on/off
* deploymentLease
* storageLease
* source catalog: if cloned from vCD library item
* source template: if cloned from vCD library item
* metadata: through the Kubernetes resources annotations: ReadOnly on vCloud Director side.

The operator also populates a `status.backing` dictionnary with the following properties according to the vCloud
Director data:

* UUID
* status
* vcd_vapp_href
* vcd_vdc_href
* owner
* deploymentLeaseInSeconds
* storageLeaseInSeconds
* metadata

If a deviation is detected with the declared `specs` of the object: a reconciliation is made to apply the state from
the declared `specs`.

> It may remains some use-case where the reconciliation will fail like when trying to apply a
power-on expected state on an *expired* object.*

The same reconciliation process will occurs when a change is made to the `specs` object declaration to apply changes
to the *backing* object on vCloud Director.

> Modification of some `specs` properties will be ignored, such as `org`, `vdc`, `fence_mode` or `accept_all_eulas`.


tests/__init__.py
"""Unit test package for kvcd."""


tests/test_kvcd.py
#!/usr/bin/env python

"""Tests for `kvcd` package."""


import unittest

from kvcd import kvcd


class TestKvcd(unittest.TestCase):
    """Tests for `kvcd` package."""

    def setUp(self):
        """Set up test fixtures, if any."""

    def tearDown(self):
        """Tear down test fixtures, if any."""

    def test_000_something(self):
        """Test something."""


kvcd/main.py
"""This module is the one to run with `kopf run` command to start the operator.

Warning: It contains a global named `get_vcd_session`, populated after the project is started.
This is probably something to cleanup in the future.
"""

import kopf
import time
import logging
import time
from dotenv import load_dotenv, find_dotenv
from kvcd.vmware.vcloud_helper import VcdSession
from kvcd.utils import setInterval
from kvcd.config import KvcdConfig


logger = logging.getLogger(__name__)


vcd_session = None
def get_vcd_session():
    """Return the current version of `vcd_session`

    Returns:
        VcdSession: current version of `vcd_session`
    """
    return vcd_session


# load dotenv file
try:
    load_dotenv(find_dotenv(raise_error_if_not_found=True, usecwd=True))
    logger.info("Configuration loaded from .env file")
except IOError:
    # Assume that the configuration is already loaded in current env.
    pass
# parse configuration from env
kvcd_config = KvcdConfig.from_environ()
logger.info("Configuration is loaded")


@kopf.on.startup()
def startup_kvcd(logger, **kwargs):
    """Startup function: create the vCD session
    """
    refresh_vcdsession()
    logger.info("vCD session is now ready")


@setInterval(sec=kvcd_config.vcd.refresh_session_interval)
def refresh_vcdsession():
    """Refresh the vCD session

    This function is run on a regular basis to update `vcd_session` with
    a working pyvcloud client.
    """
    # create the vcd session
    global vcd_session
    if isinstance(vcd_session, VcdSession):
        logger.debug("Refreshing the vCD session")
        vcd_session.rehydrate()
    else:
        logger.debug("Creating a fresh new vCD session")
        vcd_session = VcdSession(
            hostname=kvcd_config.vcd.host,
            port=kvcd_config.vcd.port,
            username=kvcd_config.vcd.username,
            password=kvcd_config.vcd.password,
            organisation=kvcd_config.vcd.org,
            verify_ssl=kvcd_config.vcd.verify_ssl,
        )


# Import the resources management functions
from kvcd.vmware.vcloud_vapp import create_vcdvapp
from kvcd.vmware.vcloud_vapp import delete_vcdvapp
from kvcd.vmware.vcloud_vapp import update_vcdvapp_description
from kvcd.vmware.vcloud_vapp import update_vcdvapp_power_state
from kvcd.vmware.vcloud_vapp import update_vcdvapp_owner
from kvcd.vmware.vcloud_vapp import update_vcdvapp_lease_info
from kvcd.vmware.vcloud_vapp import update_vcdvapp_metadata
from kvcd.vmware.vcloud_vapp import refresh_vcdvapp

kvcd/utils.py
"""Some small tools that can be used in kvcd modules
"""

import yaml
import functools
import functools
import threading


def str2bool(v:str):
    """Transform a string to a boolean value.

    Args:
        v (str): String value to convert

    Returns:
        bool: True if v is 'True', False otherwise.
    """
    return bool(yaml.safe_load(v))


def lowercase_first_string_letter(v:str):
    """Transform a string to lowercase first letter.

    Args:
        v (str): String value to convert

    Returns:
        str: Lowercase first letter of v
    """
    return v[0].lower() + v[1:]


def setInterval(sec: int):
    """Time-interval based decorator

    usage: @setInterval(sec=3)
    """
    def decorator(func):
        @functools.wraps(func)
        def wrapper(*argv, **kw):
            func(*argv, **kw)
            t = threading.Timer(sec, wrapper, argv, kw)
            t.start()
        return wrapper
    return decorator


kvcd/vmware/vcloud_helper.py
"""Set of helpers to manage a Cloud Director connection and its related objects.
"""

import atexit
import ssl
import sys
import logging
from enum import Enum

# Extra packages
from pyvcloud.vcd.client import BasicLoginCredentials
from pyvcloud.vcd.client import Client as vCDClient
from pyvcloud.vcd.client import EntityType
from pyvcloud.vcd.client import QueryResultFormat
from pyvcloud.vcd.client import ResourceType
from pyvcloud.vcd.client import MetadataDomain
from pyvcloud.vcd.client import MetadataValueType
from pyvcloud.vcd.client import MetadataVisibility
from pyvcloud.vcd.client import TaskStatus
from pyvcloud.vcd.org import Org
from pyvcloud.vcd.vapp import VApp
from pyvcloud.vcd.vdc import VDC
from pyvcloud.vcd.vm import VM
import requests
from lxml.objectify import ObjectifiedElement


logger = logging.getLogger(__name__)


class VcdSession:
    """Define VcdSession class to manage the Cloud Director connection and its related objects.
    """

    def __init__(self,
                 hostname: str,
                 username: str,
                 password: str,
                 organisation: str,
                 port: int = 443,
                 verify_ssl: bool = True):
        """Define VcdSession class based on input parameters

        Args:
            hostname (str): Hostname of the vCloud instance
            username (str): Username to use
            password (str): User's password
            organisation (str): Name of the organisation
            verify_ssl (bool, optional): Verify the vCloud SSL certificate. Defaults to True.

        Raises:
            VCDError: Any vCloud director related error.
        """
        logger.info(f'Initializing a Cloud Director session to {hostname} in organisation {organisation}')
        self._creds = BasicLoginCredentials(username, organisation, password)
        try:
            self.client = vCDClient(uri=f"https://{hostname}:{port}",
                                    verify_ssl_certs=verify_ssl,
                                    log_file=None,
                                    log_requests=False,
                                    log_headers=False,
                                    log_bodies=False)
            self.client.set_credentials(self._creds)
            atexit.register(self.__close)
        except Exception as err:
            raise VCDError(f'Unable to create the Cloud Director session: {err}')
        # shortcuts to usefull settings
        self.hostname = hostname
        self.org = Org(self.client,
                       resource=self.client.get_org())
        logger.debug(f'Connected to {self.client.get_api_uri()})')


    def rehydrate(self):
        """Renew the authentication, based on stored credentials.
        """
        self.client.set_credentials(self._creds)


    def __close(self):
        """Exit method to cloture a connection
        """
        logger.info("Closing the Cloud Director session")
        self.client.logout()
        logger.debug("Cloud Director session closed")



class VCDError(Exception):
    """Base class for exceptions with logging.
    """
    def __init__(self, msg='', *args,**kwargs):
        logger.error(f"VCDError: {msg}")
        sys.exit(msg)

    def __str__(self):
        return self.msg


def get_org(vcd_session:VcdSession, org_name: str):
    """Get an Org VDC based on its name

    Args:
        vcd_session (VcdSession): VCD session
        org_name (str): Name of the Organization

    Returns:
        Org: Org object
    """
    try:
        org_resource = vcd_session.client.get_org_by_name(org_name)
    except EntityNotFoundException:
        raise kopf.PermanentError(f"No Org found with name: {org_name}")
    org = Org(vcd_session.client, resource=org_resource)
    logger.debug(f"Org found: {org_resource.get('name')}")
    return org


def get_vdc(vcd_session: VcdSession, org_name: str, vdc_name: str):
    """Get an Org VDC based on its name

    Args:
        vcd_session (VcdSession): VCD session
        org_name (str): Name of the Organization
        vdc_name (str): Name of the VDC to get

    Returns:
        VDC: VDC object
    """
    org = get_org(vcd_session=vcd_session, org_name=org_name)
    vdc_resource = org.get_vdc(vdc_name)
    if vdc_resource == None:  # Compare to None as record.__repr()__ return an empty str: ''
        raise kopf.PermanentError(f"No Org VDC found with name: {vdc_name}")
    vdc = VDC(vcd_session.client, resource=vdc_resource)
    logger.debug(f"Org VDC found: {vdc_resource.get('name')}")
    return vdc

kvcd/config.py
"""This Submodules contains the definition of the expected configuration
to setup to use kvcd.

The main configuration is handled by `environ-config` module.
"""

import environ
import logging

logger = logging.getLogger(__name__)

@environ.config(prefix="KVCD")
class KvcdConfig:
    """kvcd configuration
    """
    @environ.config
    class VcloudConfig:
        """vcloud configuration
        """
        host = environ.var(
            help="Hostname of the vCloud instance")
        port = environ.var(
            default=443,
            help="Port of the vCloud instance",
            converter=int)
        org = environ.var(
            default="System",
            help="Organization of the vCloud instance")
        username = environ.var(
            default="Administrator",
            help="Username of the vCloud instance")
        password = environ.var(
            help="Password of the vCloud instance user")
        verify_ssl = environ.bool_var(
            default=True,
            help="Verify SSL certificate of the vCloud instance")
        refresh_session_interval = environ.var(
            default=3600,
            help="Interval (in secs) between to refresh of the authentication session",
            converter=int)

    vcd = environ.group(
        VcloudConfig,
        optional=False)
    refresh_interval = environ.var(
        default=60,
        help="Refresh interval of the vCloud instance data for each object",
        converter=int)
    refresh_initial_delay = environ.var(
        default=60,
        help="Warming up duration",
        converter=int)
    refresh_idle_delay = environ.var(
        default=60,
        help="Reduce the number of timer checks when the ressource is changed",
        converter=int)
